### DISCUSSION TITLE
Backup API

### DISCUSSION THREAD

**Discussion by kalinstaykov** [pocketbase/pocketbase#337] [Q&A] [answered]
Are there any plans to make a simple backup API that zips the `pb_data` directory and make it available for download?

That would help with hosting on places such as render.com or others where often is not easy to access the disk and copy files over directly.

**Comment by ganigeorgiev** [comment, marked as answer]
Unfortunately we cannot just zip the `pb_data` directory while the application is running (or more precisely, when there is an active write operation) since this could result in data loss (or in the worst case scenario - a sqlite db file corruption).

We can start a single large transaction and lock any writes during the backup period, but depending on your data size this may not be very performant (similar to the concerns in import/export of records data https://github.com/pocketbase/pocketbase/issues/48).

Another thing that also need to be taken in consideration is when S3 compatible storage is used (do we download and backup the uploaded files from there or only the local db state?).

This feature exists in my todo, but I plan to spend more time on this after v0.7.0-v0.8.0 release. I'll create later an issue and will add it to the [roadmap](https://github.com/orgs/pocketbase/projects/2/views/1) so that it can be tracked.

**Reply by ganigeorgiev** [reply]
Here is the related issue - https://github.com/pocketbase/pocketbase/issues/338 (It is added to the [roadmap](https://github.com/orgs/pocketbase/projects/2/views/1), somewhere after the planned documentation improvements).

**Comment by kalinstaykov** [comment]
Thank you for the quick reply! I'm probably a bit off with those but here are some thoughts on the issues you've mentioned:

* Maybe we can initiate API lock for write operations during the database backup. That will make write requests fail but it will guarantee backup consistency. That's better than copying the directory from outside the app because then we won't know if a write operation is happening and we have no control of the internal workflow.
* For big databases there is not much we can do unfortunately. It will end up being a design decision during production environment design of how that's handled.
* S3 storage should probably not be considered during DB backup. It usually has its own snapshot capabilities that would probably be better in many cases.

I saw that there are some plans to support other databases which might complicate the backup API design. Maybe external databases can be left out of this solution since database management is usually handled separately anyway. That way user can take the easy way out of using the embedded (sqlite3) database or go to the more complex setup of having external database.

**Reply by manelephant** [reply]
Hi everyone,

I would not implement backup inside the app, it is not the app responsability to create archives of a database.

There is already solutions to do this.

**Reply by ganigeorgiev** [reply]
@manelephant To some extend I agree with you, but I also think it would be nice to have a very basic rudimentary backup/snapshot option from the Admin UI, especially before importing a new collection schema.

**Comment by paulocoutinhox** [comment]
To backup:
`sqlite3 source.db ".timeout 10000" ".backup backup.db"`

Things related:
- https://www.sqlite.org/backup.html
- https://stackoverflow.com/questions/66909180/increase-the-lock-timeout-with-sqlite-and-what-is-the-default-values

**Comment by pnwmatt** [comment]
For future travelers to this thread: [sqlite_rsync](https://sqlite.org/rsync.html) is a precompiled script from [sqlite.org/downloads.html](https://sqlite.org/downloads.html) (download the correct "tools" binary)

It can replicate a running sqlite db without blocking reads or writes and is the preferred way to backup these days.
